Parser/pegen.c

```diff
@@ -2359,41 +2360,45 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
             } else {
                 unicode_string_found = 1;
             }
-            n_elements++;
+            if (!prev_is_constant) {
+                n_elements++;
+            }
         } else {
             n_elements += asdl_seq_LEN(elem->v.JoinedStr.values);
             f_string_found = 1;
         }
+        prev_is_constant = elem->kind == Constant_kind;
     }
 
     if ((unicode_string_found || f_string_found) && bytes_found) {
         RAISE_SYNTAX_ERROR("cannot mix bytes and nonbytes literals");
         return NULL;
     }
 
-    if (bytes_found) {
-        PyObject* res = PyBytes_FromString("");
-       for (i = 0; i < len; i++) {
-            expr_ty elem = asdl_seq_GET(strings, i);
-            PyBytes_Concat(&res, elem->v.Constant.value);
-        }
-        if (_PyArena_AddPyObject(arena, res) < 0) {
-            return NULL;
-        }
-        return _PyAST_Constant(res, NULL, lineno, col_offset, end_lineno, end_col_offset, p->arena);
-    }
-
     asdl_expr_seq* values = _Py_asdl_expr_seq_new(n_elements, p->arena);
     if (values == NULL) {
         return NULL;
     }
 
     Py_ssize_t current_pos = 0;
     Py_ssize_t j = 0;
+    expr_ty prev_elem = NULL;
     for (i = 0; i < len; i++) {
         expr_ty elem = asdl_seq_GET(strings, i);
         if (elem->kind == Constant_kind) {
-            asdl_seq_SET(values, current_pos++, elem);
+            if (prev_elem && prev_elem->kind == Constant_kind) {
+                /* fold consequent constants */
+                PyObject *new_str = PySequence_Concat(prev_elem->v.Constant.value,
```

Comment:
This is O(n**2). We need a PyUnicode_Writer and a PyBytes_Writer here to do this efficiency

Reply:
Check the previous implementation in `string_parser.c` for reference.

Reply:
We could:

* Get this to work using specialized APIs like `PyUnicode_ConcatAndDel` and friends.
* Try to use the optimized Bytes and Unicode writers to do this efficiency.

I would recommend doing (1) and optimize later, but `PySecuence_Concat` is too generic unfortunately